### PR TITLE
RemoteOperation Callback iOS

### PR DESCRIPTION
**Pull request by MehrdadMzfr**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [ ] Did you add unit tests if this modifies the Windows code?
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Added the callback for RemoteOperationCompletion in iOS

### CODE REVIEW

File: SDK/AppCenterData/Microsoft.AppCenter.Data.iOS.Bindings/ApiDefinition.cs [hunk 1]
```diff
@@ -201,6 +223,19 @@ interface MSWriteOptions : MSBaseOptions
...
+    [Protocol, Model]
+    [BaseType(typeof(NSObject))]
+    interface MSRemoteOperationDelegate
+    {
+        // @optional - (void)data:(MSData *)data
+        //               didCompleteRemoteOperation:(NSString*) operation
+        //                      forDocumentMetadata:(MSDocumentMetadata* _Nullable) documentMetadata
+        //                                withError:(MSDataError* _Nullable) error;
+        [Export("data:didCompleteRemoteOperation:forDocumentMetadata:withError:")]
+        void DataDidCompletedRemoteOperation(MSData data, string operation, MSDocumentMetadata documentMetadata, MSDataError error);
```
Comment (guperrot) [inline SDK/AppCenterData/Microsoft.AppCenter.Data.iOS.Bindings/ApiDefinition.cs]: need to rename to DataDidCompleteRemoteOperation

File: SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/RemoteOperationDelegate.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+namespace Microsoft.AppCenter.Data.iOS
+{
+    public class RemoteOperationDelegate : MSRemoteOperationDelegate
+    {
+        public override void DataDidCompletedRemoteOperation(MSData data, string operation, MSDocumentMetadata documentMetadata, MSDataError error)
+        {
+            OnDataDidCompletedRemoteOperationAction.Invoke(operation, documentMetadata, error);
+        }
+
+        public Action<string, MSDocumentMetadata, MSDataError> OnDataDidCompletedRemoteOperationAction { get; set; }
```
Comment (guperrot) [inline SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/RemoteOperationDelegate.cs]: If that was an inner class of Data like Android you wouldn't need the action to forward

Comment (MehrdadMzfr) [reply SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/RemoteOperationDelegate.cs]: it's good practice to keep them split, this also aligns with the Objective-C implementation

File: SDK/AppCenterData/Microsoft.AppCenter.Data.iOS.Bindings/ApiDefinition.cs (hunk 1 above)
Comment (guperrot) [reply SDK/AppCenterData/Microsoft.AppCenter.Data.iOS.Bindings/ApiDefinition.cs]: marked as resolved, but no new commit

Comment (MehrdadMzfr) [reply SDK/AppCenterData/Microsoft.AppCenter.Data.iOS.Bindings/ApiDefinition.cs]: pushed now

File: SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/Data.cs [hunk 3]
```diff
@@ -20,6 +21,25 @@ public partial class Data : AppCenterService
...
+            _remoteOperationDelegate.DataDidCompleteRemoteOperation = (operation, documentMetadata, error) =>
+            {
+                if (RemoteOperationCompleted == null)
+                {
+                    return;
+                }
+                RemoteOperationCompleted(typeof(Data), new RemoteOperationCompletedEventArgs
+                {
+                    Operation = operation,
+                    DocumentMetadata = documentMetadata.ToDocumentMetadata(),
```
Comment (achocron) [inline SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/Data.cs]: can `documentMetadata` be `null`?

File: SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/Data.cs [hunk 4]
```diff
@@ -20,6 +21,25 @@ public partial class Data : AppCenterService
...
+
+        static Data()
+        {
+            _remoteOperationDelegate.DataDidCompleteRemoteOperation = (operation, documentMetadata, error) =>
+            {
+                if (RemoteOperationCompleted == null)
+                {
+                    return;
+                }
+                RemoteOperationCompleted(typeof(Data), new RemoteOperationCompletedEventArgs
```
Comment (achocron) [inline SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/Data.cs]: Is there a possible race condition here? Maybe the `RemoteOperationCompleted` callback can be made `null` before it is invoked but after checking. Maybe save it to a local variable and then check for null/invoke?

Comment (achocron) [reply SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/Data.cs]: Or just use `RemoteOperationCompleted?.Invoke()`

**Comment by guperrot** [comment]
```
"/Users/vsts/agent/2.155.1/work/1/s/AppCenter-SDK-Build-Mac.sln" (Build target) (1) ->
"/Users/vsts/agent/2.155.1/work/1/s/SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/Microsoft.AppCenter.Data.iOS.csproj" (default target) (38) ->
(CoreCompile target) -> 
  Data.cs(28,13): error CS1656: Cannot assign to 'DataDidCompleteRemoteOperation' because it is a 'method group' [/Users/vsts/agent/2.155.1/work/1/s/SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/Microsoft.AppCenter.Data.iOS.csproj]
```

File: SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/Data.cs [hunk 5]
```diff
@@ -25,16 +25,12 @@ public partial class Data : AppCenterService
...
-                if (RemoteOperationCompleted == null)
-                {
-                    return;
-                }
-                RemoteOperationCompleted(typeof(Data), new RemoteOperationCompletedEventArgs
+                RemoteOperationCompleted?.Invoke(null, new RemoteOperationCompletedEventArgs
                 {
                     Operation = operation,
-                    DocumentMetadata = documentMetadata.ToDocumentMetadata(),
+                    DocumentMetadata = documentMetadata?.ToDocumentMetadata(),
```
Comment (guperrot) [inline SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/Data.cs]: I was looking at Android code and it does not have the null check.
What scenario causes metadata to be null? Is that real?

Comment (MehrdadMzfr) [reply SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/Data.cs]: metadata won't be null, null check wouldn't hurt, is there a way to enforce non null arguments?

Comment (guperrot) [reply SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/Data.cs]: if it's not null, we usually don't put `?.`, that's confusing.

File: SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/Data.cs [hunk 6]
```diff
@@ -20,6 +21,21 @@ public partial class Data : AppCenterService
...
+        static RemoteOperationDelegate _remoteOperationDelegate = new RemoteOperationDelegate();
+
+        static Data()
+        {
+            _remoteOperationDelegate.OnDataDidCompleteRemoteOperationAction = (operation, documentMetadata, error) =>
+            {
+                RemoteOperationCompleted?.Invoke(null, new RemoteOperationCompletedEventArgs
+                {
+                    Operation = operation,
+                    DocumentMetadata = documentMetadata?.ToDocumentMetadata(),
```
Comment (guperrot) [inline SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/Data.cs]:
```suggestion
                    DocumentMetadata = documentMetadata.ToDocumentMetadata(),
```

**Pull request merged**